server/service/transaction_service.rs

```diff
@@ -397,139 +404,161 @@ impl TransactionService {
         req_id: Uuid,
         open_req: typedb_protocol::transaction::open::Req,
     ) -> Result<ControlFlow<(), ()>, Status> {
-        let receive_time = Instant::now();
-        self.network_latency_millis = Some(open_req.network_latency_millis);
-        let mut transaction_options = TransactionOptions::default();
-        if let Some(options) = open_req.options {
-            // transaction options
-            if let Some(parallel) = options.parallel {
-                transaction_options.parallel = parallel;
-            }
-            if let Some(timeout) = options.schema_lock_acquire_timeout_millis {
-                transaction_options.schema_lock_acquire_timeout_millis = timeout
-            }
-
-            // service options
-            self.prefetch_size = options.prefetch_size.or(Some(DEFAULT_PREFETCH_SIZE));
-            self.transaction_timeout_millis =
-                options.transaction_timeout_millis.or(Some(DEFAULT_TRANSACTION_TIMEOUT_MILLIS));
-        }
-
-        let transaction_type = typedb_protocol::transaction::Type::try_from(open_req.r#type)
-            .map_err(|_| ProtocolError::UnrecognisedTransactionType { enum_variant: open_req.r#type }.into_status())?;
+        run_with_diagnostics_async(self.diagnostics_manager.clone(), ActionKind::TransactionOpen, || async {
```

Comment:
I wrap every network call-based logic in `run_with_diagnostics`(`_async`). Inside this function, we save information about actions (success/fail), and also extract the error codes when possible.

Reply:
These huge closures are a bit awkward. Can we keep the original methods and call them in `run_with_diagnostics_async` instead?

Reply:
That's less relevant for the main typedb service.

Reply:
Both methods will be private, so it will be quite easy to accidentally use the one without diagnostics. That's why I decided to just make a closure. Do you have a naming idea of how to make the intended usage clear?

Reply:
Fine, just for `TransactionService`. Thanks